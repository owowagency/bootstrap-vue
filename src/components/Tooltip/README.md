# Tooltip

```vue
<Tooltip :fallback-placements="['bottom']">
    Insert coin here
</Tooltip>
```

## Imported Props

| Prop name          | Description                                | Type                                                              | Values                                     | Default                            |
| ------------------ | ------------------------------------------ | ----------------------------------------------------------------- | ------------------------------------------ | ---------------------------------- |
| fallbackPlacements | The fallback placements of the component   | [TooltipFallbackPlacement](../../composables/useTooltipPlacement) | ['top'\|'bottom'\|'left'\|'right']         | ['top', 'right', 'bottom', 'left'] |
| placement          | The default placement of the component     | [TooltipPlacement](../../composables/useTooltipPlacement)         | ['auto'\|'top'\|'bottom'\|'left'\|'right'] | 'top'                              |
| trigger            | The trigger on which to toggle the trigger | [Trigger](../../composables/useTrigger)                           | ['click'\|'hover'\|'focus'\|'manual']      | ['hover', 'focus']                 |

<!-- TODO Auto-generate based on type? -->

## Events

| Name     | Args | Description                                                                                             |
| -------- | ---- | ------------------------------------------------------------------------------------------------------- |
| show     |      | This event fires immediately when the `show` instance method is called                                  |
| shown    |      | This event is fired when the popover has been made visible to the user                                  |
| hide     |      | This event is fired immediately when the `hide` instance method has been called                         |
| hidden   |      | This event is fired when the popover has finished being hidden from the user                            |
| inserted |      | This event is fired after the show.bs.tooltip event when the tooltip template has been added to the DOM |

## Props

| Prop name    | Description                                          | Type                             | Values | Default           |
| ------------ | ---------------------------------------------------- | -------------------------------- | ------ | ----------------- |
| boundary     | Overflow constraint boundary of the tooltip          | string \| HTMLElement            | -      | 'clippingParents' |
| container    | Appends the tooltip to the given element             | string \| HTMLElement \| boolean | -      | false             |
| customClass  | The classes that will be forwarded to the tooltip    | string\|array\|object            | -      | ''                |
| delay        | Delay showing and hiding the tooltip in milliseconds | number\|object                   | -      | 0                 |
| html         | Defines if HTML is allowed in the tooltip            | boolean                          | -      | false             |
| offset       | The x, y offset of the tooltip                       | string \| number[]               | -      | () =&gt; [0, 0]   |
| popperConfig | Change Bootstrap's default Popper config             | object\|func                     | -      | () =&gt; ({})     |
| tag          | The tag to be used for the list group item           | string                           | -      | 'div'             |
| title        | The text to display in the tooltip                   | string\|number                   | -      | ''                |

## Slots

| Name    | Description                         | Bindings |
| ------- | ----------------------------------- | -------- |
| default | Displays the content of the tooltip |          |
