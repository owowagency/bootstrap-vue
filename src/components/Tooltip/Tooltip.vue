<template>
    <Component
        :is="tag"
        ref="tooltip"
    >
        <!-- @slot Displays the content of the tooltip -->
        <slot />
    </Component>
</template>

<script lang="ts">
import {PropType, computed, onBeforeUnmount, ref} from 'vue';
import {tooltipFallbackPlacementProps, tooltipPlacementProps} from '@/composables/useTooltipPlacement';
import {triggerProps} from '@/composables/useTrigger';
import useBootstrapEmits from '@/composables/useBootstrapEmits';
import useBootstrapInstance from '@/composables/useBootstrapInstance';

const tooltipEvents = ['show', 'shown', 'hide', 'hidden', 'inserted'] as const;

type TooltipEvent = typeof tooltipEvents[number];
</script>

<script lang="ts" setup>
const props = defineProps({
    /**
     * Overflow constraint boundary of the tooltip
     */
    boundary: {
        type: [String, Object] as PropType<string | HTMLElement>,
        default: 'clippingParents',
    },
    /**
     * Appends the tooltip to the given element
     */
    container: {
        type: [String, Object, Boolean] as PropType<string | HTMLElement | boolean>,
        default: false,
    },
    /**
     * The classes that will be forwarded to the tooltip
     */
    customClass: {
        type: [String, Array, Object],
        default: '',
    },
    /**
     * Delay showing and hiding the tooltip in milliseconds
     */
    delay: {
        type: [Number, Object],
        default: 0,
    },
    /**
     * Defines if HTML is allowed in the tooltip
     */
    html: {
        type: Boolean,
        default: false,
    },
    /**
     * The x, y offset of the tooltip
     */
    offset: {
        type: [String, Array] as PropType<string | number[]>,
        default: () => [0, 0],
    },
    /**
     * Change Bootstrap's default Popper config
     */
    popperConfig: {
        type: [Object, Function],
        default: () => ({}),
    },
    /**
     * The tag to be used for the list group item
     */
    tag: {
        type: String,
        default: 'div',
    },
    /**
     * The text to display in the tooltip
     */
    title: {
        type: [String, Number],
        default: '',
    },
    ...tooltipFallbackPlacementProps,
    ...tooltipPlacementProps,
    ...triggerProps,
});

const tooltip = ref<HTMLElement>();

const emit = defineEmits<{(event: TooltipEvent): void}>();

const tooltipClass = computed(() => {
    if (typeof props.customClass === 'string') {
        return props.customClass;
    }

    if (Array.isArray(props.customClass)) {
        return props.customClass.join(' ');
    }

    return Object.keys(props.customClass)
        .filter((key: string) => !!props.customClass[key])
        .join(' ');
});

const tooltipTriggers = computed(() => {
    if (Array.isArray(props.trigger)) {
        return props.trigger.join(' ');
    }

    return props.trigger;
});

useBootstrapEmits(
    tooltip,
    tooltipEvents,
    emit,
    'tooltip',
);

const {bsInstance: bsTooltip} = useBootstrapInstance(
    'Tooltip',
    tooltip,
    {
        ...props,
        customClass: tooltipClass.value,
        trigger: tooltipTriggers.value,
    },
);

onBeforeUnmount(() => {
    bsTooltip.value?.dispose();
});

defineExpose({bsTooltip});
</script>

<docs>
```vue
<Tooltip :fallback-placements="['bottom']">
    Insert coin here
</Tooltip>
```

## Imported Props

| Prop name          | Description                                | Type                                                              | Values                                     | Default                            |
| ------------------ | ------------------------------------------ | ----------------------------------------------------------------- | ------------------------------------------ | ---------------------------------- |
| fallbackPlacements | The fallback placements of the component   | [TooltipFallbackPlacement](../../composables/useTooltipPlacement) | ['top'\|'bottom'\|'left'\|'right']         | ['top', 'right', 'bottom', 'left'] |
| placement          | The default placement of the component     | [TooltipPlacement](../../composables/useTooltipPlacement)         | ['auto'\|'top'\|'bottom'\|'left'\|'right'] | 'top'                              |
| trigger            | The trigger on which to toggle the trigger | [Trigger](../../composables/useTrigger)                           | ['click'\|'hover'\|'focus'\|'manual']      | ['hover', 'focus']                 |

<!-- TODO Auto-generate based on type? -->
## Events

| Name     | Args | Description                                                                                             |
| -------- | ---- | ------------------------------------------------------------------------------------------------------- |
| show     |      | This event fires immediately when the `show` instance method is called                                  |
| shown    |      | This event is fired when the popover has been made visible to the user                                  |
| hide     |      | This event is fired immediately when the `hide` instance method has been called                         |
| hidden   |      | This event is fired when the popover has finished being hidden from the user                            |
| inserted |      | This event is fired after the show.bs.tooltip event when the tooltip template has been added to the DOM |
</docs>
