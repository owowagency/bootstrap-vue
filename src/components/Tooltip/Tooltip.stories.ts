import {tooltipFallbackPlacements, tooltipPlacements} from '@/composables/useTooltipPlacement';
import Btn from '@/components/Button';
import Tooltip from '.';
import {triggers} from '@/composables/useTrigger';

export default {
    title: 'Components/Tooltip',
    component: Tooltip,
    argTypes: {
        fallbackPlacements: {
            control: {type: 'check'},
            options: tooltipFallbackPlacements,
        },
        placement: {
            control: {type: 'select'},
            options: tooltipPlacements,
        },
        tag: {
            control: {type: 'text'},
        },
        title: {
            control: {type: 'text'},
        },
        trigger: {
            control: {type: 'check'},
            options: triggers,
        },
    },
};

const template = (args) => ({
    components: {Btn, Tooltip},
    setup: () => ({args}),
    template: `
        <div class="d-flex justify-content-center align-items-center position-fixed top-0 bottom-0 start-0 end-0">
            <Tooltip v-bind="args" class="btn btn-primary">
                Hover me
            </Tooltip>
        </div>
    `,
});

export const Default = template.bind({});

Default.args = {
    fallbackPlacements: ['top', 'right', 'bottom', 'left'],
    placement: 'top',
    tag: 'button',
    title: 'I am the tooltip title',
    trigger: ['hover', 'focus'],
};
