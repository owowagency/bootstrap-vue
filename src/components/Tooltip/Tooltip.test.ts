import Tooltip from '@/components/Tooltip';
import {describe} from 'vitest';

describe('template', () => {
    componentRenderTest(Tooltip);

    componentWrapperTagTest(Tooltip, {props: {tag: 'button'}});
});
