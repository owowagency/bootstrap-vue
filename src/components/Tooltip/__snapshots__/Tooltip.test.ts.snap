// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > renders default 1`] = `
"<div>
  <!-- @slot Displays the content of the tooltip -->
</div>"
`;
